ibis/expr/api.py

```diff
@@ -4193,22 +4193,39 @@ def mutate(table, exprs=None, **mutations):
         for name, expr in sorted(mutations.items(), key=operator.itemgetter(0))
     )
 
+    selection_exprs_by_name = {}
     for expr in exprs:
-        if expr.get_name() and isinstance(expr, ir.DestructColumn):
-            raise com.ExpressionError(
-                f"Cannot name a destruct column: {expr.get_name()}"
-            )
+        if isinstance(expr, ir.DestructColumn):
+            if expr.get_name():
+                raise com.ExpressionError(
+                    f"Cannot name a destruct column: {expr.get_name()}"
+                )
+            struct_type = expr.type()
+            for name in struct_type.names:
+                selection_exprs_by_name[name] = expr
+        elif isinstance(expr, ir.ValueExpr):
+            selection_exprs_by_name[expr.get_name()] = expr
 
-    by_name = collections.OrderedDict(
-        (expr.get_name(), expr) for expr in exprs
-    )
     columns = table.columns
-    used = by_name.keys() & columns
+    overwriting_cols_to_expr = {
+        name: expr
+        for name, expr in selection_exprs_by_name.items()
+        if name in columns
+    }
+    non_overwriting_exprs = [
```

Comment:
We can't do a simple 'if expr not in overwriting_cols_to_expr.values()' because ibis complains with 'the truth value of an expr is not defined' type of error.